app/uk/gov/hmrc/mdgstub/controllers/MdgStubController.scala

```diff
@@ -16,76 +16,125 @@
+
+  private val availabilityMode = raw"(\d{3}):(\d+):(\d+)".r
 
   def requestTransfer() = Action.async(parse.raw) { implicit request =>
 
-    val xml = new String(request.body.asBytes().get.toArray)
+    validateXml(request.body) match {
+      case Success(xml) =>
 
-    validXml(xml) match {
-      case Success(_) =>
-        if (checkIfSimulatedFailure(xml)) {
-          Future.successful(InternalServerError("Simulated failure"))
-        } else {
-          Future.successful(NoContent)
+        withActiveAvailabilityMode(xml) {
+          case Right(Some(AvailabilityMode(status,delay,_))) if (delay.length > 0) => Promise.timeout(Status(status.toInt), delay)
+          case Right(Some(AvailabilityMode(status,_,_)))     => Future.successful((Status(status.toInt)))
+          case Right(None)                                   => Future.successful(NoContent)
+          case Left(error)                                   => Future.successful(BadRequest(error))
         }
+
       case Failure(error) =>
         Future.successful(BadRequest(error.getMessage))
     }
   }
 
-
-
-  private def validXml(body: String): Try[Unit] = {
+  private def validateXml(rawBuffer: RawBuffer): Try[String] = {
 
     val schemaLang = javax.xml.XMLConstants.W3C_XML_SCHEMA_NS_URI
+
     val xsdStream =
       new javax.xml.transform.stream.StreamSource(this.getClass.getResourceAsStream("/mdg-schema.xsd"))
+
     val schema = javax.xml.validation.SchemaFactory.newInstance(schemaLang).newSchema(xsdStream)
 
     val factory = javax.xml.parsers.SAXParserFactory.newInstance()
     factory.setNamespaceAware(true)
     factory.setSchema(schema)
 
     val validatingParser = factory.newSAXParser()
+
     val xmlLoader = new scala.xml.factory.XMLLoader[scala.xml.Elem] {
       override def parser = validatingParser
+
       override def adapter =
         new scala.xml.parsing.NoBindingFactoryAdapter {
-
-          override def error(e: SAXParseException): Unit =
-            throw e
-
+          override def error(e: SAXParseException): Unit = throw e
         }
-
     }
 
-    Try(xmlLoader.load(new StringReader(body)))
-
+    Try {
+      val xml = new String(rawBuffer.asBytes().get.toArray)
+      xmlLoader.load(new StringReader(xml)).toString
+    }
   }
 
-  private def checkIfSimulatedFailure(body : String): Boolean = {
-    val parsedXml = scala.xml.XML.loadString(body)
+  private def withActiveAvailabilityMode(xmlBody: String)(block: Either[String,Option[AvailabilityMode]] => Future[Result]): Future[Result] = {
+    val parsedXml = scala.xml.XML.loadString(xmlBody)
     val properties = for {
       property: Node <- parsedXml \ "properties" \ "property"
       name <- property \ "name"
       value <- property \ "value"
     } yield (name.text, value.text)
 
-    properties.contains(("SHOULD_FAIL", "true"))
+    val availabilityMaybe: Option[(String, String)] = properties.find {
```

Comment:
This bit is a bit confusing for me. What these true/false mean?
Why then we take _2 from it?

Reply:
Maybe:
val availabilityMaybe : Option[String] = properties.collect { case ("AVAILABILITY", mode) => mode}